src/vellum/workflows/events/types.py

```diff
@@ -57,39 +65,41 @@ class BaseDeploymentParentContext(BaseParentContext):
 
 
 class WorkflowDeploymentParentContext(BaseDeploymentParentContext):
-    type: Literal["WORKFLOW_RELEASE_TAG"] = "WORKFLOW_RELEASE_TAG"
+    type: Literal["WORKFLOW_RELEASE_TAG"]
     workflow_version_id: UUID
 
 
 class PromptDeploymentParentContext(BaseDeploymentParentContext):
-    type: Literal["PROMPT_RELEASE_TAG"] = "PROMPT_RELEASE_TAG"
+    type: Literal["PROMPT_RELEASE_TAG"]
     prompt_version_id: UUID
 
 
 class NodeParentContext(BaseParentContext):
-    type: Literal["WORKFLOW_NODE"] = "WORKFLOW_NODE"
-    node_definition: Type["BaseNode"]
-
-    @field_serializer("node_definition")
-    def serialize_node_definition(self, definition: Type, _info: Any) -> Dict[str, Any]:
-        return serialize_type_encoder(definition)
+    type: Literal["WORKFLOW_NODE"]
+    node_definition: VellumCodeResourceDefinition
 
 
 class WorkflowParentContext(BaseParentContext):
-    type: Literal["WORKFLOW"] = "WORKFLOW"
-    workflow_definition: Type["BaseWorkflow"]
-
-    @field_serializer("workflow_definition")
-    def serialize_workflow_definition(self, definition: Type, _info: Any) -> Dict[str, Any]:
-        return serialize_type_encoder(definition)
+    type: Literal["WORKFLOW"]
+    workflow_definition: VellumCodeResourceDefinition
+
+
+# Define the discriminated union
+ParentContextUnion = Annotated[
+    Union[
+        WorkflowParentContext,
+        NodeParentContext,
+        WorkflowDeploymentParentContext,
+        PromptDeploymentParentContext,
+    ],
+    Field(discriminator="type"),
+]
 
+# Create the final type
+ParentContext = Optional[ParentContextUnion]
 
```

Comment:
catch here is that parentcontext isn't quite a "type" now but a Field. but should be able to be added to models now and have type discrimination

Reply:
https://docs.pydantic.dev/latest/concepts/unions/#nested-discriminated-unions

Reply:
[nit] I would remove this here and rename the above back to `ParentContext`, and use the `Optional` in the consumers.

The link you commented here is about _multiple_ discriminators, which this data model does not yet exhibit

Reply:
after seeing that we do `Optional[ParentContext]` below and in other consumers anyway, I'm doubling down on this feedback

Reply:
🫡  should be up